packages/adblocker-extended-selectors/src/eval.ts

```diff
@@ -71,104 +71,247 @@ export function matches(element: Element, selector: AST): boolean {
+          return false;
+        }
+        value = attribute.value;
+      } else {
+        // matching attribute name by string
+        value = element.getAttribute(namePattern);
+        // null means the attribute is not present
+        if (value === null) {
+          return false;
+        }
+      }
+
+      // early exit if no value pattern is provided
+      if (!valuePattern) {
+        return true;
+      }
+
+      // wrapping quotes are optional
+      if (
+        (valuePattern.startsWith('"') && valuePattern.endsWith('"')) ||
+        (valuePattern.startsWith("'") && valuePattern.endsWith("'"))
+      ) {
+        valuePattern = valuePattern.slice(1, -1);
+      }
+
+      if (valuePattern.startsWith('/') && valuePattern.endsWith('/')) {
+        // matching value by regex
+        const regex = new RegExp(valuePattern.slice(1, -1));
+        return regex.test(value);
+      } else {
+        // matching value by string
+        return value === valuePattern;
+      }
+    } else if (selector.name === 'upward') {
+      // :upward is handled in querySelectorAll
+      return false;
     }
   }
 
   return false;
 }
 
-export function querySelectorAll(element: Element, selector: AST): Element[] {
-  const elements: Element[] = [];
+function findAncestorByDistance(element: Element, distance: number): Element | null {
+  // sanity check, since the distance is provided as part of filter
+  if (distance <= 0 || distance >= 256) {
+    return null;
+  }
+  let ancestor: Element | null = element;
+  for (let i = 0; i < distance; i++) {
+    ancestor = ancestor.parentElement;
+    if (ancestor === null) {
+      return null;
+    }
+  }
+  return ancestor;
+}
 
-  if (
-    selector.type === 'id' ||
-    selector.type === 'class' ||
-    selector.type === 'type' ||
-    selector.type === 'attribute'
-  ) {
-    elements.push(...element.querySelectorAll(selector.content));
-  } else if (selector.type === 'list') {
-    for (const subSelector of selector.list) {
-      elements.push(...querySelectorAll(element, subSelector));
+function findAncestorBySelector(element: Element, selector: string): Element | null {
+  let ancestor: Element | null = element.parentElement;
+  while (ancestor !== null) {
+    if (ancestor.matches && ancestor.matches(selector)) {
+      return ancestor;
     }
-  } else if (selector.type === 'compound') {
-    // TODO - handling compound needs to be reworked...
-    // .cls:upward(1) for example will not work with this implementation.
-    // :upward is not about selecting, but transforming a set of nodes (i.e.
-    // uBO's transpose method).
-    if (selector.compound.length !== 0) {
-      elements.push(
-        ...querySelectorAll(element, selector.compound[0]).filter((e) =>
-          selector.compound.slice(1).every((s) => matches(e, s)),
-        ),
-      );
+    ancestor = ancestor.parentElement;
+  }
+  return null;
+}
+
+function handleCompoundSelector(element: Element, compound: AST[]): Element[] {
```

Comment:
This function is so ambiguous to me. `:upward` is compound selector but also `:has` and other selectors are still chainable which means they're also compound. We should call this `handleUpward` or something. Also, this shouldn't be a function that handles only `:upward` specially.

I found `:has` is still chainable and handled in `match` with `querySelectorAll` circular structure. Shouldn't we follow that style or refactor to avoid circular structure?

Reply:
I agree. As I understand this function, it is mixing the aspect of iterating over compound (extended) CSS selectors (including, but not limited to `:upward`). And it has two paths: one where upward exists and one where it does not exist.

Separating these two concerns (visiting and matching a single compound selector) should make it more readable.